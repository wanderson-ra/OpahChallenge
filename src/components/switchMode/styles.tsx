import { isIphoneX } from "react-native-iphone-x-helper";
import { widthPercentageToDP as width, heightPercentageToDP as height } from "react-native-responsive-screen";
import SwitchSelector from "react-native-switch-selector";

import styled from "styled-components/native";

export const Container = styled.SafeAreaView`
    position: absolute;
    top: ${isIphoneX() ? height(7) : height(5)}px;
    right: ${width(3)}px;
    justify-content: center;
    align-items: center;
    flex: 1;
`;

export const Selector = styled(SwitchSelector)`
    width: ${width(30)}px;
`;
