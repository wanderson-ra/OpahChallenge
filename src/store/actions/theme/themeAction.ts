import { AnyAction } from "redux";
import { ThunkDispatch } from "redux-thunk";
import { DefaultTheme } from "styled-components/native";

import * as GetStorageUseCase from "src/useCases/getDataStorageUseCase";
import * as SaveStorageUseCase from "src/useCases/saveDataStorageUseCase";

import colorTheme from "src/configurations/themes/theme";

import { MODE_THEME_KEY } from "src/globals/constants/storage";

export type Mode = "light" | "dark";

import { ThemeActionTypes } from "../types";

export interface ThemeAction {
    type: ThemeActionTypes.MODE_THEME;
    mode: Mode;
    theme: DefaultTheme;
}

export type Action = ThemeAction;

export const dispatchThemeAction = (mode: Mode, theme: DefaultTheme): ThemeAction => {
    return {
        type: ThemeActionTypes.MODE_THEME,
        mode: mode,
        theme: theme,
    };
};

export const saveMode = (mode: Mode) => {
    return async (dispatch: ThunkDispatch<void, void, AnyAction>): Promise<void> => {
        await SaveStorageUseCase.save(MODE_THEME_KEY, mode);

        const theme = mode === "dark" ? colorTheme.dark : colorTheme.light;

        dispatch(dispatchThemeAction(mode, theme));
    };
};

export const getMode = () => {
    return async (dispatch: ThunkDispatch<void, void, AnyAction>): Promise<void> => {
        const mode = await GetStorageUseCase.get(MODE_THEME_KEY);

        if (mode) {
            const theme = mode === "dark" ? colorTheme.dark : colorTheme.light;
            dispatch(dispatchThemeAction(mode === "light" ? "light" : "dark", theme));
        } else {
            dispatch(dispatchThemeAction("light", colorTheme.light));
        }
    };
};
