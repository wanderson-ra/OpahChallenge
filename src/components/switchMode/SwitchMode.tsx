import React, { useCallback } from "react";
import { isIphoneX } from "react-native-iphone-x-helper";
import { heightPercentageToDP as height } from "react-native-responsive-screen";
import SwitchSelector from "react-native-switch-selector";
import { useDispatch, useSelector } from "react-redux";

import { useTheme } from "styled-components/native";

import { saveMode } from "src/store/actions/theme/themeAction";
import { Mode } from "src/store/actions/theme/themeAction";
import { RootState } from "src/store/reducers";

import { Container, Selector } from "./styles";

export const SwitchMode: React.FC = () => {
    const { mode } = useSelector((state: RootState) => state.themeReducer.themeReducer);

    const theme = useTheme();

    const dispatch = useDispatch();

    const save = useCallback(
        (mode: Mode): void => {
            dispatch(saveMode(mode));
        },
        [dispatch]
    );

    return (
        <Container>
            <Selector
                as={SwitchSelector}
                backgroundColor={theme.switch.backgroundColor}
                height={isIphoneX() ? height(3.5) : height(4)}
                initial={mode === "light" ? 0 : 1}
                onPress={(value: string): void => save(value === "light" ? "light" : "dark")}
                textColor={theme.font.primary}
                selectedColor={theme.font.inverse}
                buttonColor={theme.switch.color}
                borderColor={theme.card.background}
                hasPadding
                options={[
                    { label: "light", value: "light" },
                    { label: "dark", value: "dark" },
                ]}
            />
        </Container>
    );
};
