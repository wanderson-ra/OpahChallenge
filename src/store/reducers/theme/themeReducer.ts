import { combineReducers } from "redux";
import { DefaultTheme } from "styled-components/native";

import { Action, Mode } from "../../actions/theme/themeAction";
import { ThemeActionTypes } from "../../actions/types";

export interface ThemeReducer {
    mode: Mode | undefined;
    theme: DefaultTheme | undefined;
}

export interface ThemeState {
    themeReducer: ThemeReducer;
}

export const themeReducer = (
    state: ThemeReducer = {
        mode: undefined,
        theme: undefined,
    },
    action: Action
): ThemeReducer => {
    switch (action.type) {
        case ThemeActionTypes.MODE_THEME:
            return {
                ...state,
                mode: action.mode,
                theme: action.theme,
            };

        default:
            return state;
    }
};

export default combineReducers<ThemeState>({
    themeReducer,
});
