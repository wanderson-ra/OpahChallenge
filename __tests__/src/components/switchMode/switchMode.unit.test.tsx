import React from "react";
import SwitchSelector from "react-native-switch-selector";
import { Provider } from "react-redux";
import * as Redux from "react-redux";
import renderer from "react-test-renderer";

import { when } from "jest-when";
import { Store } from "redux";
import configureStore from "redux-mock-store";
import { ThemeProvider } from "styled-components/native";

import colorTheme from "../../../../src/configurations/themes/theme";

import * as ThemeAction from "../../../../src/store/actions/theme/themeAction";
import { ThemeState, ThemeReducer } from "../../../../src/store/reducers/theme/themeReducer";

import { SwitchMode } from "../../../../src/components/switchMode/SwitchMode";

const mockStore = configureStore([]);

describe("Test of SwitchMode", () => {
    let store: Store;

    let switchMode: renderer.ReactTestRenderer;

    it("Test save mode light", async () => {
        const mode = "dark";
        const expectMode = "light";

        const mockThemeReducer: ThemeReducer = {
            mode: mode,
            theme: colorTheme.light,
        };

        const initialPropsWith: ThemeState = {
            themeReducer: mockThemeReducer,
        };

        store = mockStore({
            userReducer: initialPropsWith,
        });

        store.dispatch = jest.fn();

        const mockedUseSelector = jest.spyOn(Redux, "useSelector");
        when(mockedUseSelector).calledWith(expect.any).mockReturnValue(mockThemeReducer);

        const mockedThemeAction = jest.spyOn(ThemeAction, "saveMode");

        await renderer.act(async () => {
            switchMode = renderer.create(
                <Provider store={store}>
                    <ThemeProvider theme={colorTheme.dark}>
                        <SwitchMode />
                    </ThemeProvider>
                </Provider>
            );
        });

        const switchSelector = switchMode.root.findByType(SwitchSelector);

        await renderer.act(async () => {
            switchSelector.props.onPress(expectMode);
        });

        expect(mockedThemeAction).toBeCalledWith(expectMode);
    });

    it("Test snapshot", async () => {
        const mode = "light";

        const mockThemeReducer: ThemeReducer = {
            mode: mode,
            theme: colorTheme.light,
        };

        const initialPropsWith: ThemeState = {
            themeReducer: mockThemeReducer,
        };

        store = mockStore({
            userReducer: initialPropsWith,
        });

        store.dispatch = jest.fn();

        const mockedUseSelector = jest.spyOn(Redux, "useSelector");
        when(mockedUseSelector).calledWith(expect.any).mockReturnValue(mockThemeReducer);

        await renderer.act(async () => {
            switchMode = renderer.create(
                <Provider store={store}>
                    <ThemeProvider theme={colorTheme.dark}>
                        <SwitchMode />
                    </ThemeProvider>
                </Provider>
            );
        });

        expect(switchMode.toJSON()).toMatchSnapshot();
    });
});
