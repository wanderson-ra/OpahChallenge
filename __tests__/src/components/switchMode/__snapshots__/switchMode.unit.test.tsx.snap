// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test of SwitchMode Test snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "justifyContent": "center",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
        },
        Array [
          Object {
            "width": 225,
          },
        ],
      ]
    }
  >
    <View
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        onLayout={[Function]}
        style={
          Object {
            "backgroundColor": "#1d1d1d",
            "borderRadius": 50,
            "height": 53.5,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "borderColor": "#1d1d1d",
              "borderRadius": 50,
              "borderWidth": 1,
              "flex": 1,
              "flexDirection": "row",
            }
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "center",
                "opacity": 1,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "backgroundColor": "transparent",
                    "color": "#000000",
                    "fontSize": 14,
                    "fontWeight": "normal",
                    "textAlign": "center",
                  },
                  Object {},
                ]
              }
            >
              light
            </Text>
          </View>
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "center",
                "opacity": 1,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "backgroundColor": "transparent",
                    "color": "#FFFFFF",
                    "fontSize": 14,
                    "fontWeight": "normal",
                    "textAlign": "center",
                  },
                  Object {},
                ]
              }
            >
              dark
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
